hydromt/models/components/model_component.py

```diff
@@ -0,0 +1,22 @@
+"""Abstract base class for ModelComponent."""
+
+from abc import ABC, abstractmethod
+
+
+class ModelComponent(ABC):
+    """Abstract base class for ModelComponent."""
+
+    @abstractmethod
+    def set(self):
+        """Set data."""
+        pass
+
+    @abstractmethod
+    def write(self):
+        """Write data."""
+        pass
+
+    @abstractmethod
+    def read(self):
+        """Read data."""
+        pass
```

Comment:
All these functions don't seem to be implemented by GridComponent at all, because they have different parameters. We should consider how we want to implement them.

We either need to go for a dynamic solution with `**kwargs`, or make an `initialize(self, foo, bar, baz)` that is not in the base class ModelComponent, or we make all setters per property: `GridComponent.model_root = X`, `GridComponent.region = Y`.

Reply:
I think we need to add more functions here especially the __init__ with a weakref to model. Suggestion from @DirkEilander  in PR #810 

@deltamarnix: is there an issue with using abstractmethod for methods that the subclasses MUST have (here set, read, write are musts).
We can check afterwards when merging together with ModelRoot if we need extra properties as we have yet some unresolved issues there before we can check.

``` python
class ModelComponent:

    def __init__(self, model):
        self._model_ref = weakref.ref(model)

    @property
    def model(self) -> hydromt.Model:
        # Access the Model instance through the weak reference
        return self._model_ref()
    
    @property
    def data_catalog(self) -> hydromt.DataCatalog:
        return self.model.data_catalog
    
    @property
    def logger(self) -> logging.Logger:
        return self.model.logger

    @property
    def model_root(self) -> str:
        return self.model.root
```